### PR TITLE
Disable account access for unverified accouts

### PR DESCRIPTION
**Pull request by mkangia**
## Product Description

Disable access to ~"My Account"~ "Two Factor" setup section for a user if they don't have any active domains.
This is related to a security concern flagged during an audit, which blocks a user from creating an account using their email address, in case there is already an account created with their email address and has 2 factor auth setup.

Steps:
1. Someone creates an account on HQ using another user's email address, say barackobama@gov.in
2. they setup 2 factor auth on this account
3. they can't verify the account so they can't really use it
4. the original user, Barack, wants to sign up on HQ but sees a message "Username already linked with another account". Even if he would reset the password through forgot password, he would still not be able to access the account due to 2 factor auth.
Thus, Barack is totally blocked on having an account on HQ using their own email address.

## Technical Summary

When a new user confirms their account, their [project space is marked as active](https://github.com/dimagi/commcare-hq/blob/caa619713ae566d4baab1debfea9f764d34410a2/corehq/apps/registration/views.py#L516).
For a user invited, from a domain, user is added to the domain they were invited from when they accept invite sent on their email.
So, it seems that the way to know if a user is verified, is by checking if they have active domains.
I added a decorator to the base view for account section so it disables access to all it's children views for an unverified user.
Domain specific views are protected by `login_and_domain_required` decorator, though non-domain specific views would need additional check, for example, [quick_find](https://github.com/dimagi/commcare-hq/blob/c8313096ea0d9b69ae3a2a4e1541047f4868ca4f/corehq/apps/hqwebapp/views.py#L1204) currently just needs user to be logged in and then would rely on the following view to handle checks for access, which is okay though a check on quick_find itself would be better.

### Safety story

Tested locally

### Rollback instructions

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Comment by mkangia** [comment]
@djmore9 
Hello,
Can you confirm this change in behavior is okay with Product?

**Comment by esoergel** [comment]
> So, it seems that the way to know if a user is verified, is by checking if they have active domains.

I'm not certain of the mechanism, but I believe it's common for web users to not have their own domain.  There's a `user.is_active` property - I thought this captured this process to some extent?

**Comment by mkangia** [comment]
Hey @esoergel 

> but I believe it's common for web users to not have their own domain

[This is how we get domains for a web user](https://github.com/dimagi/commcare-hq/blob/843ca3c34d07ebeaeb839f535a3cc0008e7d9cda/corehq/apps/users/models.py#L2438)
So, active domains would include domains they own or domains they are a member of.
If a users signs up on HQ, a project space is automatically created for them, or if they are invited, they are added to the domain they were invited from.
Is there any way a user can make an account on HQ but not have a domain associated with them?

For "is_active", a new user is active [as default](https://github.com/dimagi/commcare-hq/blob/d58c94383e9883ba67c7e6355e9cd213f14617a5/corehq/apps/registration/utils.py#L98), as done from the `ProcessRegistrationView` on signup. What we need here is a user whose email has been verified, so, even an unverified user would be active.

**Comment by esoergel** [comment]
> For "is_active", a new user is active [as default](https://github.com/dimagi/commcare-hq/blob/d58c94383e9883ba67c7e6355e9cd213f14617a5/corehq/apps/registration/utils.py#L98), as done from the ProcessRegistrationView on signup. What we need here is a user whose email has been verified, so, even an unverified user would be active.

That's frustrating.  Still, I don't love the idea of using a proxy metric for this.  Essentially what this says is "if this user is a member of any active domains, they must have verified their email address".  I do think that is true (eg, if you're invited to an existing domain, you have to be able to access the invitation email, which confirms that you can access that email), but it seems indirect.

In an ideal world, we'd set a flag on the user when they first follow a link that was sent to their email address, whether a domain invitation or project confirmation.  We won't have historical access to that data though, so even if we added it, we'd have to use a proxy to back-populate it anyways.   In my opinion, it would be good to run this fix by the security team specifically.

**Comment by mkangia** [comment]
Hey @esoergel @millerdev 

I checked with @ctsims about this at length, and his take was positive for this change, https://dimagi.slack.com/archives/C0J60R34J/p1683297934045779?thread_ts=1683019340.547419&cid=C0J60R34J

Just checking if this makes you positive about this change and we can go ahead with it?

(there is a test failure which looks related to how the test it setup, which I will fix once the solution proposed by the PR gets an approval)

**Review by esoergel — APPROVED** [review]
Thanks for the follow-up, Manish!  I think the caveats have been discussed sufficiently, and this needn't be blocked.

**Comment by ctsims** [comment]
@millerdev Unfortunately this feels very hard to feature flag since it's new-signup behavior. We could in theory make it server configurable but I'm more worried about the split in that case.

I think the "signup for 2FA via email" approach is clever and would support that pattern in general, but I think that introduces a whole new paradigm for how 2FA works which I don't think is fair to attach to this work as a meets minimum. 

> > I don't see a real scenario where a legit user wont have access to My Accounts section after this change.
> 
> This seems hard to audit.

I think this is the core underlying concern, and from my perspective the question here isn't technical but workflow driven, so it does feel fair to provide more clarity. 

IE: Manish is re-using the existing HQ logic and conditions for "is the account activated yet," which already broadly affects the user experience, not introducing a new one. If a user would have tripped this logic today, they would already be seeing a big "Activate your account" screen on login. 

I think the core question is whether a user in that state is more blocked after this change than before it, and it does feel fair to try and understand and resolve that question. 

The ways that someone would end up blocked would be if they

- Could currently create a project space, but wouldn't be able to after the change
- Could currently accept a project space invite, but wouldn't be able to after the change
- Could reasonably want to perform some set of actions in My Account, but wouldn't be able to after the change.

The last one is probably the most reasonable to ask, and specifically if this change would mean a non-active user couldn't **delete** their account that seems non-ideal unless we think there's a separate flow for that (like: "when account activations/invites expire we delete the stub account + when a user signs up for a new account with an existing stub that is a clean user experience). 

@mkangia - One thing I'll mention on this is that if India Division's goal is the minimum viable changeset, I think the smallest and cleanest change would probably be to disable **only** 2FA provisioning as a consequence of this flag, and not the rest of My Account. I do think it's a net positive to the system to have a cleaner experience around signup and to not see account actions, but there are a few points folks have brought up towards those goals that I think are bigger work lifts that ID can contribute here, so minimizing the footprint of impact might be the best approach until someone can do a deeper dive.

**Comment by mkangia** [comment]
thanks for jumping in @ctsims 

> I think the smallest and cleanest change would probably be to disable only 2FA provisioning 

That works for me.
Going to update to block access to Two Factor listing page and two factor setup page only.

**Comment by esoergel** [comment]
@mkangia yeah that is weird - I don't see any obvious references to it even in the PR where it was added:
https://github.com/dimagi/commcare-hq/pull/10010

Only idea I have is that maybe that's a URL pattern followed by the `two_factor` library - like maybe it assembles that URL path from the base two factor view.

**Review by esoergel — APPROVED** [review]

**Comment by mkangia** [comment]
Thank you all.

With scope limited to just 2 factor page, I have decided NOT to take this through QA. Tested locally that all is well.
Updated the PR description accordingly as well.

Will leave it here for a day before merging in case anyone has any thoughts/concerns.

**Pull request merged**
